File: examples/python/client_schedules.py
"""ekoDB Python Client - Schedule Management Integration Example

Demonstrates the full schedule lifecycle:
  create -> list -> get -> update -> trigger -> pause -> resume -> delete

The example creates a temporary no-op function for the schedule to invoke.
"""

import asyncio
import os

from ekodb_client import Client, Stage


async def main():
    print("=== ekoDB Schedule Management Example (Python) ===\n")

    base_url = os.getenv("API_BASE_URL", "http://localhost:8080")
    api_key = os.getenv("API_BASE_KEY")
    if not api_key:
        raise ValueError("API_BASE_KEY environment variable is required")

    client = Client.new(base_url, api_key)

    function_label = f"schedule_noop_python_{os.getpid()}"
    function_id = await client.save_function(
        {
            "label": function_label,
            "name": "Schedule example no-op",
            "parameters": {},
            "functions": [Stage.return_response({"triggered": True}, 200)],
        }
    )

    schedule_id = None
    try:
        print("--- create_schedule ---")
        schedule = await client.create_schedule(
            {
                "name": "Nightly Report",
                "function_label": function_label,
                "cron_expression": "0 0 0 * * *",
                "description": "Generate and email nightly analytics report",
                "timezone": "UTC",
                "enabled": True,
            }
        )
        schedule_id = schedule["id"]
        print(f"Created schedule: {schedule.get('name')} (id: {schedule_id})")

        print("\n--- list_schedules ---")
        print(f"Schedules: {await client.list_schedules()}")

        print("\n--- get_schedule ---")
        fetched = await client.get_schedule(schedule_id)
        print(f"Fetched: {fetched.get('name')} cron={fetched.get('cron_expression')}")

        print("\n--- update_schedule ---")
        updated = await client.update_schedule(
            schedule_id,
            {
                "cron_expression": "0 30 1 * * *",
                "description": "Changed to 1:30 AM UTC",
            },
        )
        print(f"Updated cron: {updated.get('cron_expression')}")

        print("\n--- trigger_schedule ---")
        print(f"Triggered: {await client.trigger_schedule(schedule_id)}")

        print("\n--- pause_schedule ---")
        print(
            f"Paused: enabled={(await client.pause_schedule(schedule_id)).get('enabled')}"
        )

        print("\n--- resume_schedule ---")
        print(
            f"Resumed: enabled={(await client.resume_schedule(schedule_id)).get('enabled')}"
        )
    finally:
        cleanup_error = None
        if schedule_id is not None:
            try:
                await client.delete_schedule(schedule_id)
                print("Schedule deleted successfully")
            except Exception as exc:
                cleanup_error = exc
        try:
            await client.delete_function(function_id)
        except Exception as exc:
            if cleanup_error is None:
                cleanup_error = exc
        if cleanup_error is not None:
            raise cleanup_error

    print("\n=== Schedule management example completed ===")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/go/client_schedules.go
// ekoDB Go Client - Schedules Integration Example
//
// Exercises the full schedule lifecycle:
//   CreateSchedule -> ListSchedules -> GetSchedule -> UpdateSchedule
//   -> TriggerSchedule -> PauseSchedule -> ResumeSchedule -> DeleteSchedule
//
// A temporary no-op function makes the trigger step self-contained.

package main

import (
	"errors"
	"fmt"
	"log"
	"os"

	ekodb "github.com/ekoDB/ekodb-client-go"
	"github.com/joho/godotenv"
)

func run() (runErr error) {
	fmt.Print("=== ekoDB Schedules Integration Example (Go) ===\n\n")

	_ = godotenv.Load()

	baseURL := os.Getenv("API_BASE_URL")
	if baseURL == "" {
		baseURL = "http://localhost:8080"
	}
	apiKey := os.Getenv("API_BASE_KEY")
	if apiKey == "" {
		return errors.New("API_BASE_KEY environment variable is required")
	}

	client, err := ekodb.NewClient(baseURL, apiKey)
	if err != nil {
		return fmt.Errorf("create client: %w", err)
	}

	functionLabel := fmt.Sprintf("schedule_noop_go_%d", os.Getpid())
	functionID, err := client.SaveFunction(ekodb.UserFunction{
		Label:      functionLabel,
		Name:       "Schedule example no-op",
		Parameters: map[string]ekodb.ParameterDefinition{},
		Functions: []ekodb.FunctionStageConfig{
			ekodb.StageReturn(map[string]interface{}{"triggered": true}, 200),
		},
	})
	if err != nil {
		return fmt.Errorf("save function: %w", err)
	}
	defer func() {
		if functionID != "" {
			if cleanupErr := client.DeleteFunction(functionID); cleanupErr != nil {
				runErr = errors.Join(runErr, fmt.Errorf("delete function %s: %w", functionID, cleanupErr))
			}
		}
	}()
	var scheduleID string
	defer func() {
		if scheduleID != "" {
			if cleanupErr := client.DeleteSchedule(scheduleID); cleanupErr != nil {
				runErr = errors.Join(runErr, fmt.Errorf("delete schedule %s: %w", scheduleID, cleanupErr))
			}
		}
	}()

	// 1. Create a schedule
	fmt.Println("--- Creating schedule ---")
	schedule, err := client.CreateSchedule(map[string]interface{}{
		"name":            "nightly-backup",
		"description":     "Runs a database backup every night at midnight",
		"function_label":  functionLabel,
		"cron_expression": "0 0 0 * * *",
		"enabled":         true,
	})
	if err != nil {
		return fmt.Errorf("create schedule: %w", err)
	}
	scheduleID = ekodb.GetStringValue(schedule["id"])
	if scheduleID == "" {
		return errors.New("create schedule response contained no ID")
	}
	fmt.Printf("Created schedule: %s (id: %s)\n", schedule["name"], scheduleID)

	// 2. List all schedules
	fmt.Println("\n--- Listing schedules ---")
	list, err := client.ListSchedules()
	if err != nil {
		return fmt.Errorf("list schedules: %w", err)
	}
	fmt.Printf("Schedules: %v\n", list)

	// 3. Get schedule by ID
	fmt.Println("\n--- Getting schedule ---")
	fetched, err := client.GetSchedule(scheduleID)
	if err != nil {
		return fmt.Errorf("get schedule: %w", err)
	}
	if ekodb.GetStringValue(fetched["name"]) != "nightly-backup" || ekodb.GetStringValue(fetched["cron_expression"]) != "0 0 0 * * *" {
		return fmt.Errorf("fetched schedule had unexpected name/cron: %v", fetched)
	}
	fmt.Printf("Fetched schedule: %s (cron: %s)\n", fetched["name"], fetched["cron_expression"])

	// 4. Update schedule
	fmt.Println("\n--- Updating schedule ---")
	updated, err := client.UpdateSchedule(scheduleID, map[string]interface{}{
		"cron_expression": "0 0 2 * * *",
		"description":     "Runs backup at 2 AM instead of midnight",
	})
	if err != nil {
		return fmt.Errorf("update schedule: %w", err)
	}
	if cron := ekodb.GetStringValue(updated["cron_expression"]); cron != "0 0 2 * * *" {
		return fmt.Errorf("updated schedule cron was %q", cron)
	}
	fmt.Printf("Updated cron: %v\n", updated["cron_expression"])

	// 5. Trigger immediately
	fmt.Println("\n--- Triggering schedule ---")
	triggered, err := client.TriggerSchedule(scheduleID)
	if err != nil {
		return fmt.Errorf("trigger schedule: %w", err)
	}
	if status := ekodb.GetStringValue(triggered["status"]); status != "triggered" {
		return fmt.Errorf("trigger schedule status was %q", status)
	}
	fmt.Printf("Trigger response: %v\n", triggered)

	// 6. Pause schedule
	fmt.Println("\n--- Pausing schedule ---")
	paused, err := client.PauseSchedule(scheduleID)
	if err != nil {
		return fmt.Errorf("pause schedule: %w", err)
	}
	if enabled, ok := paused["enabled"].(bool); !ok || enabled {
		return fmt.Errorf("paused schedule enabled value was %v", paused["enabled"])
	}
	fmt.Printf("Schedule enabled: %v\n", paused["enabled"])

	// 7. Resume schedule
	fmt.Println("\n--- Resuming schedule ---")
	resumed, err := client.ResumeSchedule(scheduleID)
	if err != nil {
		return fmt.Errorf("resume schedule: %w", err)
	}
	if enabled, ok := resumed["enabled"].(bool); !ok || !enabled {
		return fmt.Errorf("resumed schedule enabled value was %v", resumed["enabled"])
	}
	fmt.Printf("Schedule enabled: %v\n", resumed["enabled"])

	// 8. Delete schedule
	fmt.Println("\n--- Deleting schedule ---")
	if err := client.DeleteSchedule(scheduleID); err != nil {
		return fmt.Errorf("delete schedule: %w", err)
	}
	scheduleID = ""
	fmt.Println("Schedule deleted")
	if err := client.DeleteFunction(functionID); err != nil {
		return fmt.Errorf("delete function: %w", err)
	}
	functionID = ""

	fmt.Println("\n=== All schedule operations completed successfully ===")
	return nil
}

func main() {
	if err := run(); err != nil {
		log.Fatal(err)
	}
}
